_source/_posts/2022-05-15-jhipster-ionic-spring-boot.adoc

```diff
@@ -0,0 +1,481 @@
+    <key>CFBundleURLSchemes</key>
+    <array>
+      <string>capacitor</string>
+      <string>dev.localhost.ionic</string>
+    </array>
+  </dict>
+</array>
+----
+
+Modify the JHipster app's CORS settings (in `backend/src/main/resources/config/application-prod.yml`) to allow `capacitor://localhost` as an origin.
+
+[source,yaml]
+----
+jhipster:
+  ...
+  cors:
+    allowed-origins: 'http://localhost:8100,capacitor://localhost'
+----
+
+Restart your backend app.
+Deploy your Ionic app to iOS Simulator and run it.
+
+[source,shell]
+----
+npx cap run ios
+----
+
+Confirm you can log in and rejoice in your success!
+
+image::{% asset_path 'blog/jhipster-ionic-blueprint/flickr2-simulator.png' %}[alt=Flickr2 running on iOS,width=300,align=center]
+
+=== Run your Ionic app on Android
+
+Generate an Android project with Capacitor.
+
+[source,shell]
+----
+npx ionic capacitor add android
+----
+
+Enable clear text traffic and add `dev.localhost.ionic` as a scheme in `android/app/src/main/AndroidManifest.xml`:
+
+[source,xml]
+----
+<activity ... android:usesCleartextTraffic="true">
+  <!-- You'll need to add this intent filter so redirects work -->
+  <intent-filter>
+    <action android:name="android.intent.action.VIEW" />
+    <category android:name="android.intent.category.DEFAULT" />
+    <category android:name="android.intent.category.BROWSABLE" />
+    <data android:scheme="dev.localhost.ionic" />
+    <!--data android:scheme="com.okta.dev-133337" /-->
+  </intent-filter>
+
+  <intent-filter>
+    <action android:name="android.intent.action.MAIN" />
+    <category android:name="android.intent.category.LAUNCHER" />
+  </intent-filter>
+</activity>
+----
+
+Modify the JHipster app's CORS settings to allow `\http://localhost` as an origin.
+
+[source,yaml]
+----
+jhipster:
+  ...
+  cors:
+    allowed-origins: 'http://localhost:8100,capacitor://localhost,http://localhost'
+----
+
+Restart your backend app and run your Ionic app on Android using the Capacitor CLI:
+
+[source,shell]
+----
+npx cap run android
+----
+
+CAUTION: If you get an error when running this command, make sure you're using Java 11.
+
+You'll need to run a couple commands to allow the emulator to communicate with JHipster and Keycloak.
+
+[source,shell]
+----
+adb reverse tcp:8080 tcp:8080
+adb reverse tcp:9080 tcp:9080
+----
+
+You should be able to log in and edit entities, just like you can in a browser and on iOS!
+
+image::{% asset_path 'blog/jhipster-ionic-blueprint/flickr2-android.png' %}[alt=Flickr2 running on Android,width=300,align=center]
+
+== Why use OpenID Connect for mobile apps?
+
+Storing link:/blog/2019/01/22/oauth-api-keys-arent-safe-in-mobile-apps[API keys and secrets in mobile apps is not safe].
+OAuth 2.0 solves this problem by not shipping any secrets in mobile apps, and instead involving the user in the process of getting an access token into the app.
+These access tokens are unique per user and every time they log in.
```

Comment:
```suggestion
These access tokens are unique per user and they update every time the user logs in.
```
Without this edit, the sentence feels a little ambiguous to me. Feel free to tweak as needed.